### PR TITLE
[Feature] Flux2 klein inpaint

### PR DESCRIPTION
**Pull request by RuixiangMa**
## Purpose
Add Flux inpainting

## Test Result

CASE 0:
```shell
curl -s -X POST "http://localhost:8004/v1/images/edits" \
-F "image=@dinosaur_input.jpeg"   \
-F "prompt=Change the yellow dinosaur to green one"   \
-F "guidance_scale=1.0"   \
-F "num_inference_steps=4"   \
-F "n=1" | jq -r '.data[0].b64_json' | base64 --decode > output.png
``` 

CASE 1:
```shell
curl -s -X POST "http://localhost:8004/v1/images/edits" \
-F "image=@dinosaur_input.jpeg"   \
-F "mask_image=@dinosaur_mask.png"   \
-F "prompt=Change the yellow dinosaur to green one"   \
-F "guidance_scale=1.0"   \
-F "num_inference_steps=4"   \
-F "n=1" | jq -r '.data[0].b64_json' | base64 --decode > output.png
``` 

CASE 2:
```shell
curl -s -X POST "http://localhost:8004/v1/images/edits" \
-F "image=@the-ball-stadion-football-the-pitch.jpeg"   \
-F "mask_image=@ball_mask.png"   \
-F "reference_image=@ball_reference.jpeg"   \
-F "prompt=Replace the ball"   \
-F "guidance_scale=1.0"   \
-F "num_inference_steps=4"   \
-F "n=1" | jq -r '.data[0].b64_json' | base64 --decode > output.png
``` 

| Case | Input Image | Mask / Reference | Prompt | Result |
|:----:|:-----------:|:----------------:|:-------|:------:|
| **CASE 0** | Input | None | `Change the yellow dinosaur to green one` | Result |
| **CASE 1** | Input | Mask | `Change the yellow dinosaur to green one` | Result |
| **CASE 2** | Input | Reference<br>*(with small mask)* | `Replace the ball` | Result |

### CODE REVIEW

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 1]
```diff
@@ -1981,6 +1982,29 @@ async def _create_diffusion_chat_completion(
                 except Exception as e:  # pragma: no cover - safeguard
                     logger.warning("Failed to parse LoRA request: %s", e)
 
+            # Add mask image if provided
```
Comment (ZJY0516) [inline vllm_omni/entrypoints/openai/serving_chat.py]: Could we reuse image edit api?

Comment (RuixiangMa) [reply vllm_omni/entrypoints/openai/serving_chat.py]: > Could we reuse image edit api?

Thanks for the tip—already updated!

File: vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py [hunk 2]
```diff
@@ -856,17 +937,37 @@ def forward(
...
+        if mask_image is not None and (image is None or (isinstance(image, list) and len(image) == 0)):
+            raise ValueError("image must be provided when using mask_image for inpainting")
+
+        init_image = None
+        if condition_images:
+            first_image = condition_images[0]
+            if hasattr(first_image, "to"):
+                init_image = first_image.to(dtype=torch.float32)
+            else:
+                init_image = torch.from_numpy(np.array(first_image)).float() / 255.0
```
Comment (SamitHuang) [inline vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py]: is it possible that `condition_images` is int tensor in range [0, 255]?

File: vllm_omni/entrypoints/openai/api_server.py [hunk 3]
```diff
@@ -1039,6 +1039,11 @@ async def edit_images(
     background: str | None = Form("auto"),
     output_compression: Annotated[int, Form(ge=0, le=100)] = 100,
     user: str | None = Form(None),  # unused now
+    # vllm-omni extensions for image editing
+    mask_image: UploadFile | None = File(None),
+    mask_url: str | None = Form(None, alias="mask_url"),
```
Comment (SamitHuang) [inline vllm_omni/entrypoints/openai/api_server.py]: can we merge `mask_image` and `mask_url` into one arg for simplicity?

File: vllm_omni/entrypoints/openai/api_server.py [hunk 4]
```diff
@@ -1039,6 +1039,11 @@ async def edit_images(
...
+    reference_image: UploadFile | None = File(None),
+    reference_url: str | None = Form(None, alias="reference_url"),
```
Comment (SamitHuang) [inline vllm_omni/entrypoints/openai/api_server.py]: ditto

File: vllm_omni/entrypoints/openai/api_server.py (hunk 3 above)
Comment (RuixiangMa) [reply vllm_omni/entrypoints/openai/api_server.py]: Double checked ， preprocess() already returns a float32 tensor with values in [-1, 1], so all extra type checks and conversions are unnecessary

File: vllm_omni/entrypoints/openai/api_server.py (hunk 4 above)
Comment (RuixiangMa) [reply vllm_omni/entrypoints/openai/api_server.py]: already cleaned it up

File: vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py (hunk 2 above)
Comment (RuixiangMa) [reply vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py]: > is it possible that `condition_images` is int tensor in range [0, 255]?

Double checked ， preprocess() already returns a float32 tensor with values in [-1, 1], so all extra type checks and conversions are unnecessary

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by RuixiangMa** [comment]
cc @hsliuustc0106

**Comment by hsliuustc0106** [comment]
do we need to update the docs?

File: vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py [hunk 5]
```diff
@@ -891,6 +988,71 @@ def forward(
...
+                batch_size=batch_size * num_images_per_prompt,
+                generator=generator,
+                device=device,
+                dtype=self.vae.dtype,
+            )
+
+        if reference_image_latent_ids is not None:
+            latent_ids = torch.cat([latent_ids, reference_image_latent_ids], dim=1)
+        elif image_latent_ids is not None:
+            latent_ids = torch.cat([latent_ids, image_latent_ids], dim=1)
```
Comment (nuclearwu) [inline vllm_omni/diffusion/models/flux2_klein/pipeline_flux2_klein.py]: @RuixiangMa The code saves `original_latent_ids` at the beginning, but its purpose is unclear and it's only used once at the end.

**Code Flow:**

```python
# Early in forward() - Save original value
original_latent_ids = latent_ids  # Text prompt position IDs only

# Later - Modify latent_ids based on conditions
if reference_image_latent_ids is not None:
    latent_ids = torch.cat([latent_ids, reference_image_latent_ids], dim=1)
elif image_latent_ids is not None:
    latent_ids = torch.cat([latent_ids, image_latent_ids], dim=1)
# Now latent_ids includes image position IDs

# Throughout denoising loop - Use modified latent_ids
latent_image_ids = latent_ids  # Modified value

# At the end - Use original value for decoding
latents = self._unpack_latents_with_ids(latents, original_latent_ids)
# Why not use latent_ids (which includes image positions)?
```

**Questions:**

1. Why save `original_latent_ids` if we modify `latent_ids` throughout the function?
2. What is the intended purpose of `original_latent_ids`?
3. Why use the original (text-only) IDs for decoding when latents were generated with image conditioning?

**Comment by RuixiangMa** [comment]
> do we need to update the docs?

Done. Updated image_edit_api.md

**Comment by hsliuustc0106** [comment]
any tests for this model? I think we can use a random weight to test the offline/online for the same input, the output should be the same.

**Review by Gaohan123 — COMMENTED** [review]
Could you please add some tests? Thanks

**Comment by RuixiangMa** [comment]
@Gaohan123 @hsliuustc0106 updated tests

**OFFLINE:**

tests/e2e/offline_inference/test_flux2_klein_inpaint.py::test_flux2_klein_inpaint_basic PASSED                                       [ 33%]
tests/e2e/offline_inference/test_flux2_klein_inpaint.py::test_flux2_klein_inpaint_deterministic PASSED                               [ 66%]
tests/e2e/offline_inference/test_flux2_klein_inpaint.py::test_flux2_klein_inpaint_different_seeds_different_output PASSED            [100%]

**ONLINE:**

tests/e2e/online_serving/test_flux2_klein_inpaint_expansion.py::test_flux2_klein_inpaint_basic[tp2_basic] PASSED                                                                                                                                                         [ 33%]
tests/e2e/online_serving/test_flux2_klein_inpaint_expansion.py::test_flux2_klein_inpaint_deterministic[tp2_basic] PASSED                                                                                                                                                 [ 66%]
tests/e2e/online_serving/test_flux2_klein_inpaint_expansion.py::test_flux2_klein_inpaint_multiple_outputs[tp2_basic] PASSED                                                                                                                                              [100%]

**Review by hsliuustc0106 — APPROVED** [review]
